indexer/indexer.go

```diff
@@ -0,0 +1,383 @@
+	"github.com/bnb-chain/bsc-staking-indexer/contract/stakehub"
+	"github.com/bnb-chain/bsc-staking-indexer/model"
+	"github.com/bnb-chain/bsc-staking-indexer/store"
+	"github.com/bnb-chain/bsc-staking-indexer/util/log"
+)
+
+var (
+	RtyNum      = uint(5)
+	RtyAttempts = retry.Attempts(RtyNum)
+	RtyDelay    = retry.Delay(time.Millisecond * 50)
+	RtyErr      = retry.LastErrorOnly(true)
+
+	dialer = &net.Dialer{
+		Timeout:   time.Second,
+		KeepAlive: 60 * time.Second,
+	}
+
+	transport = &http.Transport{
+		DialContext:         dialer.DialContext,
+		MaxIdleConnsPerHost: 2000,
+		MaxConnsPerHost:     2000,
+		IdleConnTimeout:     90 * time.Second,
+		TLSClientConfig:     &tls.Config{InsecureSkipVerify: true},
+	}
+
+	httpClient = &http.Client{
+		Timeout:   5 * time.Second,
+		Transport: transport,
+	}
+)
+
+type Indexer interface {
+	Start(ctx context.Context)
+}
+
+type Config struct {
+	RPCURL                    string
+	StartBlockNumber          int64
+	NumberOfBlocksForFinality int64
+	Delegators                []common.Address
+}
+
+func New(ctx context.Context, cfg Config, store store.Store) (Indexer, error) {
+	client, err := ethclient.DialOptions(ctx, cfg.RPCURL, rpc.WithHTTPClient(httpClient))
+	if err != nil {
+		log.Errorw("indexer: failed to dial ethclient", "err", err)
+		return nil, err
+	}
+
+	cursor, err := store.QueryCursor(ctx)
+	if err != nil {
+		log.Errorw("indexer: failed to query cursor", "err", err)
+		return nil, err
+	}
+	if cfg.StartBlockNumber > cursor {
+		cursor = cfg.StartBlockNumber - 1
+	}
+
+	stakeHub, err := stakehub.New(client)
+	if err != nil {
+		log.Errorw("indexer: failed to new stakehub", "err", err)
+		return nil, err
+	}
+
+	validatorInfos, err := store.QueryValidatorInfos(ctx)
+	if err != nil {
+		log.Errorw("indexer: failed to query validator infos", "err", err)
+		return nil, err
+	}
+
+	stakeCredits := make(map[common.Address]*stakecredit.ContractWithInfo)
+	for _, v := range validatorInfos {
+		stakeCredits[v.Credit], err = stakecredit.New(v.Operator, v.Credit, client)
+		if err != nil {
+			log.Errorw("indexer: failed to new stakecredit", "err", err)
+			return nil, err
+		}
+	}
+
+	return &indexer{
+		cursor:                    cursor,
+		NumberOfBlocksForFinality: cfg.NumberOfBlocksForFinality,
+		delegators:                cfg.Delegators,
+		store:                     store,
+		client:                    client,
+		stakeHub:                  stakeHub,
+		stakeCredits:              stakeCredits,
+	}, nil
+}
+
+type indexer struct {
+	cursor                    int64
+	NumberOfBlocksForFinality int64
+	delegators                []common.Address
+	parentHeader              *types.Header
+
+	store        store.Store
+	client       *ethclient.Client
+	stakeHub     *stakehub.Contract
+	stakeCredits map[common.Address]*stakecredit.ContractWithInfo // credit address -> stake credit
+}
+
+func (i *indexer) Start(ctx context.Context) {
+	ticker := time.NewTicker(3 * time.Second)
```

Comment:
when we initial launch the indexer, because it lag a lot, there is no need to wait for each 3 seconds. 

It could be smarter: if lag a lot, no need to wait; if lag few blocks, wait for 1 second.